network/src/protocols/discovery/mod.rs

```diff
@@ -454,11 +454,7 @@ where
     let res_msg = recv_msg(substream.substream)
         .timeout(timeout)
         .map_err(Into::<NetworkError>::into)
-        .map(|r| match r {
-            Ok(Ok(msg)) => Ok(msg),
-            Ok(Err(e)) => Err(e),
-            Err(e) => Err(e),
-        })
+        .map(|r| r.and_then(|x| x))
```

Comment:
I have no clue why I didn't use `Result::and_then` when I did this conversion

Reply:
My bet: because this should properly be a flat_map and it's frustrating that it can't be.